### PR TITLE
Refresh token before using listener id

### PR DESCRIPTION
**Pull request by tetienne**
.

### CODE REVIEW

File: pyoverkiz/client.py [hunk 1]
```diff
@@ -244,8 +248,6 @@ async def somfy_tahoma_refresh_token(self, refresh_token: str) -> str:
                 seconds=token["expires_in"] - 5
             )
 
-            return self._access_token
```
Comment (tetienne) [inline pyoverkiz/client.py:247]: It was not used, as we can directly use `self._access_token`

File: pyoverkiz/client.py [hunk 2]
```diff
@@ -210,18 +211,21 @@ async def somfy_tahoma_get_access_token(self) -> str:
 
             return self._access_token
 
-    async def somfy_tahoma_refresh_token(self, refresh_token: str) -> str:
+    async def refresh_token(self) -> None:
```
Comment (tetienne) [inline pyoverkiz/client.py:214]: We don’t need to give in parameter the refresh_token, we can directly use the inner one.

File: pyoverkiz/client.py [hunk 3]
```diff
@@ -128,6 +128,7 @@ async def __aexit__(
     async def close(self) -> None:
         """Close the session."""
         if self.event_listener_id:
+            await self._refresh_token_if_expired()
```
Comment (iMicknl) [inline pyoverkiz/client.py:131]: Why is this required? Since __post calls this function as well?

Comment (tetienne) [reply pyoverkiz/client.py:131]: Indeed, this is a mistake.

File: pyoverkiz/client.py (hunk 1 above)
Comment (iMicknl) [reply pyoverkiz/client.py:247]: Indeed! Do we need to check if we have a refresh token, or can this race condition never happen.

File: pyoverkiz/client.py [hunk 4]
```diff
@@ -519,6 +521,7 @@ async def fetch_events(self) -> list[Event]:
         Per-session rate-limit : 1 calls per 1 SECONDS period for this particular
         operation (polling)
         """
+        await self._refresh_token_if_expired()
```
Comment (iMicknl) [inline pyoverkiz/client.py:523]: Why is this required here? Since we call post a line later and in post we also call this function?

File: pyoverkiz/client.py [hunk 5]
```diff
@@ -629,8 +633,8 @@ async def __get(self, path: str) -> Any:
         """Make a GET request to the OverKiz API"""
         headers = {}
 
-        if self.server == SUPPORTED_SERVERS["somfy_europe"]:
-            await self._refresh_somfy_tahoma_token_if_expired()
+        await self._refresh_token_if_expired()
```
Comment (iMicknl) [inline pyoverkiz/client.py:635]: Just a double check, here we call this before we check if there is an access token. In post we call this after we checked if we have an access token.

File: pyoverkiz/client.py (hunk 2 above)
Comment (iMicknl) [reply pyoverkiz/client.py:213]: Why the rename? I didn't like the previous one as well, however this only applies to Somfy TaHoma (Europe).

File: pyoverkiz/client.py (hunk 1 above)
Comment (tetienne) [reply pyoverkiz/client.py:247]: In our code, we have already a check: https://github.com/iMicknl/python-overkiz-api/blob/0c45f2e6a3ae20b7af77ec8cd976a393a2a3f780/pyoverkiz/client.py#L729

but not directly in this method,  I can raise an error in such case. What do you think?

File: pyoverkiz/client.py (hunk 5 above)
Comment (tetienne) [reply pyoverkiz/client.py:635]: That’s because in the `__post`, I don’t want this method to be call for the `login` endpoint. It’s not about the access_token.

File: pyoverkiz/client.py (hunk 2 above)
Comment (tetienne) [reply pyoverkiz/client.py:213]: It’s with the idea to have something generic, and also avoid to test each time the server type before calling this method.

File: pyoverkiz/client.py (hunk 1 above)
Comment (iMicknl) [reply pyoverkiz/client.py:247]: Would be good to raise an error / exception (or call the other function), just to be sure.

**Comment by iMicknl** [comment]
@tetienne will this also fix https://github.com/iMicknl/ha-tahoma/issues/736?

**Comment by tetienne** [comment]
> @tetienne will this also fix [iMicknl/ha-tahoma#736](https://github.com/iMicknl/ha-tahoma/issues/736)?

If he has been successfully logged at least once, it can fix it I guess. But if it’s from the boot, it won’t fix it.

**Comment by tetienne** [comment]
Tested locally, backoff is not anymore involved when we refresh the token, meaning we are using the correct lister id. 
The only downside is we lost all the events when we switch the listener.

**Review by iMicknl — APPROVED** [review]
Ready to merge if tested. Will this be better than current situation? Or do we need to wait for other changes?

**Comment by tetienne** [comment]
It will be better than the current situation indeed as we don't register multiple times the listener.

For the moment, the only downside as discussed, is the fact we can miss some events ever hour when we register the new listener.

I have an of fix, but I can do it in another pr.

File: pyoverkiz/client.py [hunk 6]
```diff
@@ -519,6 +523,7 @@ async def fetch_events(self) -> list[Event]:
         Per-session rate-limit : 1 calls per 1 SECONDS period for this particular
         operation (polling)
         """
+        await self._refresh_token_if_expired()
```
Comment (iMicknl) [inline pyoverkiz/client.py:526]: Why do we call this for cozytouch login? We don't use it there?

File: pyoverkiz/client.py (hunk 4 above)
Comment (tetienne) [reply pyoverkiz/client.py:526]: Because we want to refresh the `event_listener_id` before fetching the events. That’s the point of this PR.
If we don’t do that, we will call the `__post` with an expired token, and so an invalid `event_listener_id`. Even if we refresh the token into the `__post` method, the `event_listener_id` which is a parameter will still be wrong.
